src/StmtToViz.cpp

```diff
@@ -0,0 +1,2892 @@
+        // Traverse the module to populate the list of
+        // nodes we need to map and generate their assembly
+        // markers (comments that appear in the assembly code
+        // associating the code with this node)
+        for (const auto &fn : m.functions()) {
+            fn.body.accept(this);
+        }
+
+        // Find markers in asm code
+        std::istringstream asm_stream(code);
+        std::string line;
+        int lno = 1;
+        while (getline(asm_stream, line)) {
+            // Try all markers
+            std::vector<uint64_t> matched_nodes;
+            for (auto const &[node, regex] : markers) {
+                if (std::regex_search(line, regex)) {
+                    // Save line number
+                    lnos[node] = lno;
+                    // Save this node's id
+                    matched_nodes.push_back(node);
+                }
+            }
+            // We map to the first match, stop
+            // checking matched nodes
+            for (auto const &node : matched_nodes) {
+                markers.erase(node);
+            }
+
+            lno++;
+        }
+    }
+
+    int get_asm_lno(uint64_t node_id) {
+        if (lnos.count(node_id)) {
+            return lnos[node_id];
+        }
+        return -1;
+    }
+
+private:
+    // Generate asm markers for Halide loops
+    int loop_id;
+    int gen_loop_id() {
+        return ++loop_id;
+    }
+
+    std::regex gen_loop_asm_marker(int id, const std::string &loop_var) {
+        std::regex dollar("\\$");
+        std::string marker = "%\"" + std::to_string(id) + "_for " + loop_var;
+        marker = std::regex_replace(marker, dollar, "\\$");
+        return std::regex(marker);
+    }
+
+    // Generate asm markers for Halide producer/consumer ndoes
+    int prodcons_id;
+    int gen_prodcons_id() {
+        return prodcons_id;
+    }
+
+    std::regex gen_prodcons_asm_marker(int id, const std::string &var, bool is_producer) {
+        std::regex dollar("\\$");
+        std::string marker = "%\"" + std::to_string(id) + (is_producer ? "_produce " : "_consume ") + var;
+        marker = std::regex_replace(marker, dollar, "\\$");
+        return std::regex(marker);
+    }
+
+    // Mapping of IR nodes to their asm markers
+    std::map<uint64_t, std::regex> markers;
+
+    // Mapping of IR nodes to their asm line numbers
+    std::map<uint64_t, int> lnos;
+
+    using IRVisitor::visit;
+
+    void visit(const ProducerConsumer *op) override {
+        // Generate asm marker
+        markers[(uint64_t)op] = gen_prodcons_asm_marker(gen_prodcons_id(), op->name, op->is_producer);
+        // Continue traversal
+        IRVisitor::visit(op);
+    }
+
+    void visit(const For *op) override {
+        // Generate asm marker
+        markers[(uint64_t)op] = gen_loop_asm_marker(gen_loop_id(), op->name);
+        // Continue traversal
+        IRVisitor::visit(op);
+    }
+};
+
+/******************* HTMLCodePrinter Class *******************/
```

Comment:
Please use the following comment style for multi-line comments like this for uniformity with the rest of Halide and for better doxygen docs:

```
/** ClassName
 * description...
 */
```